catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/SearchFormsApplication.java

```diff
@@ -0,0 +1,239 @@
+import org.codice.ddf.catalog.ui.forms.data.FormAttributes;
+import org.codice.ddf.catalog.ui.forms.data.QueryTemplateMetacardImpl;
+import org.codice.ddf.catalog.ui.forms.data.ResultTemplateMetacardImpl;
+import org.codice.ddf.catalog.ui.forms.filter.VisitableFilterNode;
+import org.codice.ddf.catalog.ui.forms.filter.VisitableXmlElement;
+import org.codice.ddf.catalog.ui.forms.model.FilterNodeValueSerializer;
+import org.codice.ddf.catalog.ui.forms.model.JsonModel.FieldFilter;
+import org.codice.ddf.catalog.ui.forms.model.JsonModel.FormTemplate;
+import org.codice.ddf.catalog.ui.forms.model.JsonTransformVisitor;
+import org.codice.ddf.catalog.ui.util.EndpointUtil;
+import org.codice.ddf.security.common.Security;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import spark.servlet.SparkApplication;
+
+/** Provides an internal REST interface for working with custom form data for Intrigue. */
+public class SearchFormsApplication implements SparkApplication {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SearchFormsApplication.class);
+
+  private static final Security SECURITY = Security.getInstance();
+
+  private static final ObjectMapper MAPPER =
+      JsonFactory.create(
+          new JsonParserFactory().usePropertyOnly(),
+          new JsonSerializerFactory()
+              .addPropertySerializer(new FilterNodeValueSerializer())
+              .useAnnotations()
+              .includeEmpty()
+              .includeDefaultValues()
+              .setJsonFormatForDates(false));
+
+  private final CatalogFramework catalogFramework;
+
+  private final EndpointUtil util;
+
+  public SearchFormsApplication(CatalogFramework catalogFramework, EndpointUtil util) {
+    this.catalogFramework = catalogFramework;
+    this.util = util;
+  }
+
+  /**
+   * Called via blueprint on initialization. Reads configuration in {@code etc/forms} and
+   * initializes Solr with query and result templates using the {@link Metacard#TITLE} field as a
+   * unique key.
+   */
+  public void setup() {
+    Function<Map<String, Result>, Set<String>> titles =
+        m ->
+            m.entrySet()
+                .stream()
+                .map(Map.Entry::getValue)
+                .map(Result::getMetacard)
+                .map(Metacard::getTitle)
+                .collect(Collectors.toSet());
+
+    Set<String> queryTitles = queryAsAdmin(FormAttributes.Query.TAG, titles);
+    Set<String> resultTitles = queryAsAdmin(FormAttributes.Result.TAG, titles);
+    List<Metacard> initialTemplates = config().get();
+
+    initialTemplates
+        .stream()
+        .filter(QueryTemplateMetacardImpl::isQueryTemplateMetacard)
+        .filter(m -> !queryTitles.contains(m.getTitle()))
+        .forEach(this::saveMetacard);
+
+    initialTemplates
+        .stream()
+        .filter(ResultTemplateMetacardImpl::isResultTemplateMetacard)
+        .filter(m -> !resultTitles.contains(m.getTitle()))
+        .forEach(this::saveMetacard);
```

Comment:
I think you can rework the saveMetacard method to call once to the framework for all template metacards.